Create a skeleton for a new script

kazooo opened the pull request:
- add Solr util class

Commit:
Create a skeleton for a new script

 - add Solr util class

Romx98 committed:
Writting a method for all roots from solr

- added solr constant depends to this method
- filter all root uuid which dont contains CNNB or ISSN
- added wrap method for query str in SolrUtils

kazooo commented on src/main/java/cz/mzk/scripts/SyncFoxmlFieldsWithSolr.java:
```diff
@@ -61,11 +64,23 @@ public static void main(String[] args) throws ParserConfigurationException, Tran
     }
 
-    private static SolrQuery createFetchAllRootsWithoutTheRequiredFieldsSolrQuery() {
+    private static SolrQuery createFetchAllRootsWithoutTheRequiredFieldsSolrQuery(int max_rows) {
```
max_rows -> maxRows, pouzivame camelCase

kazooo replied:
modifikator *final*

kazooo commented on src/main/java/cz/mzk/scripts/SyncFoxmlFieldsWithSolr.java:
```diff
@@ -61,11 +64,23 @@ public static void main(String[] args) throws ParserConfigurationException, Tran
         // filter roots that doesn't have required fields
-        return null;
+
+        SolrQuery solrQuery = new SolrQuery();
```
*final*

kazooo commented on src/main/java/cz/mzk/utils/SolrUtils.java:
```diff
@@ -9,4 +10,13 @@
     }
+
+    public static String wrapQueryStrForEmptyValue(String fieldName) {
+        return "-" + fieldName + ":[\"\" TO * ]";
```
nerozumim co to je :) pokud chces query na prazdne pole pouzivej "-fieldName:*"

Romx98 replied:
Problem je v tom, ze v solri sa nachadza vzdy(vacsia cast) issn filed. Je tam nastavena issn:"", co pomocou -issn:* nenajdem, lebo toto predpoklada, ze tam nieje ziaden issn field.
Cize budem musiet kontrolovat, ci sa tam field issn nachadza + ci issn field je prazdy.
A tento prikaz ["" TO *], ak som dobre pochopil, riesi to, ze mi kontroluje hodnotu "" pre vsetky dokumenty.

Romx98 replied:
Napisal som ti spravu, ak budes mat cas pozrel by si sa na nu prosim. :)

kazooo commented on src/main/java/cz/mzk/utils/SolrUtils.java:
```diff
@@ -9,4 +10,13 @@
         inputDocument.addField(fieldKey, Collections.singletonMap("set", fieldValue));
     }
+
+    public static String wrapQueryStrForEmptyValue(String fieldName) {
```
mozna *queryNoFieldValue* je tady lepsi... co myslis?

Romx98 replied:
Znie to lepsie, (lepsie vystihuje)

kazooo commented on src/main/java/cz/mzk/utils/SolrUtils.java:
```diff
@@ -9,4 +10,13 @@
+        return "-" + fieldName + ":[\"\" TO * ]";
+    }
+
+    public static String wrapNegatedQueryByRegexStr(String fieldName, String searchText) {
```
mozna *queryNoStrFieldValueByRegex* je tady lepsi... co myslis?

kazooo commented on src/main/java/cz/mzk/scripts/SyncFoxmlFieldsWithSolr.java:
```diff
@@ -61,11 +64,23 @@ public static void main(String[] args) throws ParserConfigurationException, Tran
 
-    private static SolrQuery createFetchAllRootsWithoutTheRequiredFieldsSolrQuery() {
+    private static SolrQuery createFetchAllRootsWithoutTheRequiredFieldsSolrQuery(int max_rows) {
         // TODO implementation
```
komentare uz nejsou potreba, pokud uz je funkce implementovana

kazooo commented on src/main/java/cz/mzk/scripts/SyncFoxmlFieldsWithSolr.java:
```diff
@@ -61,11 +64,23 @@ public static void main(String[] args) throws ParserConfigurationException, Tran
+                SolrUtils.wrapQueryStrForEmptyValue(SolrField.ISSN) + " OR " +
+                SolrUtils.wrapNegatedQueryByRegexStr(SolrField.DC_IDENT, "cnb"));
+
+        solrQuery.addFilterQuery("{!frange l=1 u=1 v=eq(" + SolrField.UUID +", " + SolrField.ROOT_PID + ")}");
```
mozna presunout do SolrUtils, filtrovani korenu se jeste bude pouzivat v dalsich scriptech :)

kazooo reviewed (commented):
modifikator *final* je jedna z best practices, tim rikas ze je hodnota konstantni a omylem ji nikde neprepises a neudelas chyb :)

Romx98 committed:
Modify code based on cr

- renaming methods
- added new method for filter query root pid
- modified formal param to camelCase format
- modified logic of regex search method

Romx98 committed:
Parsing identifier from foxml

- added private method for parsing identifier by name
- added methods for getting identifier CNB or ISSN

Romx98 committed:
Modified code in fedora client class

- added constant DC
- added method get xml from dc datastream
- fix xpathexpression to parse identifier
- check if parse values list is empty

Romx98 committed:
Modified SolrUtils

- added final statement to formal params
- modify Query for fetch all root uuid

kazooo commented on src/main/java/cz/mzk/utils/SolrUtils.java:
```diff
@@ -11,12 +12,16 @@ public static void setModify(final SolrInputDocument inputDocument, final String
+        return "-" + fieldName + ":" + "/.*" + searchText + ".*/";
+    }
+
+    public static String filterQueryRootPid() {
```
asi filterQueryOnlyRootPids

kazooo commented on src/main/java/cz/mzk/utils/FoxmlUtils.java:
```diff
@@ -8,16 +8,60 @@
+    public Optional<List<String>> getListOfCNNBFromFoxml(final Document doc) {
+        Validate.notNull(doc);
+
+        return parseIdentifierByValueName(doc, "cnb");
```
bude dobry pokud se ti vrati jen prazdny list, optional bych rekl uz je zbytecny

kazooo commented on src/main/java/cz/mzk/utils/FoxmlUtils.java:
```diff
@@ -8,16 +8,60 @@
+    }
+
+
+    private Optional<List<String>> parseIdentifierByValueName(final Document doc, final String valueName) {
```
jen prazdny list, pokud nenajde zadny identifikator

kazooo commented on src/main/java/cz/mzk/utils/FoxmlUtils.java:
```diff
@@ -8,16 +8,60 @@
+
+        try {
+            final NodeList attrNodes = (NodeList) identifierXPath.evaluate(doc, XPathConstants.NODESET);
+            List<String> values = new ArrayList<>();
```
final

kazooo commented on src/main/java/cz/mzk/utils/FoxmlUtils.java:
```diff
@@ -8,16 +8,60 @@
+    public Optional<String> getStrISSNFromFoxml(final Document doc) {
+        Validate.notNull(doc);
+
+        Optional<List<String>> listOfISSN = parseIdentifierByValueName(doc, "issn");
```
final

kazooo commented on src/main/java/cz/mzk/utils/FoxmlUtils.java:
```diff
@@ -8,16 +8,60 @@
+            List<String> values = new ArrayList<>();
+
+            for (int i = 0; i < attrNodes.getLength(); i++) {
+                Node attrNode = attrNodes.item(i);
```
final

kazooo commented on src/main/java/cz/mzk/utils/FoxmlUtils.java:
```diff
@@ -8,16 +8,60 @@
+            }
+            return Optional.of(values);
+        } catch (XPathExpressionException e) {
+            log.warn("Can't retrieve identifier!");
```
pridej do logogvani, prosim, identifikator, ktere se nepodarilo najit, at je videt co se hledalo :)

Romx98 committed:
Modify FoxmlUtil based on PR

- modify method named in SolrUtils
- added final keyword
- removal optional class from methods returning List

kazooo commented on src/main/java/cz/mzk/utils/FoxmlUtils.java:
```diff
@@ -34,32 +35,32 @@ public FoxmlUtils() {
+            log.warn("Can't retrieve identifier \"" + valueName + "\"");
             e.printStackTrace();
-            return Optional.empty();
+            return null;
```
prazdny list

kazooo commented on src/main/java/cz/mzk/utils/FoxmlUtils.java:
```diff
@@ -34,32 +35,32 @@ public FoxmlUtils() {
-        if (listOfISSN.isPresent() && !listOfISSN.get().isEmpty()) {
-            return Optional.of(listOfISSN.get().get(0));
+        final List<String> listOfISSN = parseIdentifierByValueName(doc, "issn");
+        if (listOfISSN != null && !listOfISSN.isEmpty()) {
```
pokud funkce vrati prazdny list, tak neni potreba kontrolovat na null

Romx98 committed:
Complete scripts

- addition of synchronizer method logic
- added new method for send InputDoc to solr
- modify return value (emptyList)
- new method to get the current time (update MODIFIED_DATE)

- comments have not been deleted yet, will be deleted after final editing.

Romx98 commented:
Inak dobre by bolo, keby sa zmenene/doplnene dokumenty niekde ukladali (nejaky zoznam). Kvoli uprave tych istych dokumentov v K7.

Romx98 committed:
Added final keyword

kazooo commented on src/main/java/cz/mzk/rest/CustomSolrClient.java:
```diff
@@ -89,6 +90,18 @@ public long getNumFound(final SolrQuery queryParams) {
         return numFound;
     }
 
+    public void sendSolrInputDocument(SolrInputDocument solrInputDoc)  {
```
ve wrapperech muzes nechat nazvy metod v rozhrani stejne jako u obalovaneho objektu, tedy *add* :) ale klidne muzes nechat takhle

kazooo commented on src/main/java/cz/mzk/rest/CustomSolrClient.java:
```diff
@@ -89,6 +90,18 @@ public long getNumFound(final SolrQuery queryParams) {
+        try {
+            solrClient.add(solrInputDoc);
+        } catch (SolrServerException e) {
+            log.warn("Can't added SolrInputDocument to the custom Solr client! " + e.getMessage());
```
pro obe dve situace muzes vypsat stejny error message, neco jako "Can't index a document to the Solr instance"
pridavas dokument do instanci Solru, ne klientu :)

kazooo commented on src/main/java/cz/mzk/scripts/SyncFoxmlFieldsWithSolr.java:
```diff
@@ -42,6 +45,16 @@ public static void main(String[] args) throws ParserConfigurationException, Tran
             // apply each field synchronizer
                 // pass FOXML and SolrInputDocument object as parameters
             // if the result SolrInputDocument object has any field to update, send it to Solr
+            String uuidDoc = (String) solrDoc.getFieldValue(SolrField.UUID);
```
final

kazooo commented on src/main/java/cz/mzk/scripts/SyncFoxmlFieldsWithSolr.java:
```diff
@@ -42,6 +45,16 @@ public static void main(String[] args) throws ParserConfigurationException, Tran
+            SolrInputDocument solrInputDoc = new SolrInputDocument();
+            Optional<Document> foxml = fedoraClient.getDC(uuidDoc);
+
+            foxml.ifPresent(doc -> fieldSynchronizers.forEach(x -> x.accept(doc, solrInputDoc)));
```
pokud neexistuje, tak tady muzes pridat nejake bud logovani nebo zapis do souboru, ze dokument neexistuje ve Fedore, ale je v Solru

kazooo commented on src/main/java/cz/mzk/scripts/SyncFoxmlFieldsWithSolr.java:
```diff
@@ -53,13 +66,28 @@ public static void main(String[] args) throws ParserConfigurationException, Tran
             // add all required Solr field names and XPath expression parsing to SolrField.java and FoxmlUtils.java
             // parse the required field in FOXML
             // if exists and is not blank, update SolrInputDocument object by SolrUtils.java
+            List<String> listOfCNB = new FoxmlUtils().getListOfCNBFromFoxml(foxml);
```
final

kazooo replied:
prosim FoxmUtils jako staticky attribut scriptu, aby se pokazde nevytvarel pri volani *createCNBSynchronizer* a *createISSNSynchronizer*

kazooo commented on src/main/java/cz/mzk/scripts/SyncFoxmlFieldsWithSolr.java:
```diff
@@ -53,13 +66,28 @@ public static void main(String[] args) throws ParserConfigurationException, Tran
         return (foxml, solrInputDoc) -> {
             // TODO implementation
             // the same as in the previous function for ISSN
+            Optional<String> issn = new FoxmlUtils().getStrISSNFromFoxml(foxml);
```
final

kazooo commented on src/main/java/cz/mzk/scripts/SyncFoxmlFieldsWithSolr.java:
```diff
@@ -76,6 +104,7 @@ private static SolrQuery createFetchAllRootsWithoutTheRequiredFieldsSolrQuery(fi
         solrQuery.addField(SolrField.UUID);
         solrQuery.addField(SolrField.DC_IDENT);
         solrQuery.addField(SolrField.ISSN);
+        solrQuery.addField(SolrField.MODIFIED_DATE);
```
proc potrebujes modified_date?

kazooo commented on src/main/java/cz/mzk/utils/SolrUtils.java:
```diff
@@ -4,14 +4,21 @@
     }
 
+    public static String findOutCurrentTimeStr() {
+        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");
```
staci kdyz do modified_date zadas jen object **new Date()**, naformatovany string neni treba
myslim ze nazev metody *getCurrentTimeDate* je lepsi, co myslis?

Romx98 replied:
Pri tomto som si neni isty, ci nebude odlisne formatovanie nez v predoslich zaznamov Solri.

kazooo commented:
> Inak dobre by bolo, keby sa zmenene/doplnene dokumenty niekde ukladali (nejaky zoznam). Kvoli uprave tych istych dokumentov v K7.

mas pravdu, vytvor prosim dalsi utilitu do utils, kterou pak vytvoris jako staticky object a bude ti logovat do souboru, jehoz nazev bude obsahovat kratky popis a datum kontroly, napr neco jako "updated-issn-cnb-roots_21-11-2021-18:25:00.txt"

pak by to vypadalo najak takhle:
```java
// current timestamp do nazvu souboru object doplni samostatne
// je treba vymyslet nejaky lepsi nazev tridy :) 
private static final MyFileWritter updatedIssnCnbRootsFileWritter = new MyFileWritter("updated-issn-cnb-roots");
...
updatedIssnCnbRootsFileWritter.writeLine(rootUuid);
```

stejnym zpusobem logovat napr. i uuid, ktere nejsou ve Fedore, ale Solr je ma, ty pak zpracovavat v ramci jineho issue

kazooo commented:
jop, a jeste jedna vec, setujes nove DC_IDENT, ale Solr zaznam uz ma nejaky DC_IDENT zaindexovany, takze svymi zmeny to prepises. Tady je potreba vzit stary DC_IDENT, ktery ti prichazi ze Solru, a jenom pak ho doplnit o nove identifikatory (ISSN, CNB) a poslat zpatky. Tim se ti neztrati ostatni zaindexovane hodnoty v DC_IDENT

Romx98 commented:
```
if (!listOfCNB.isEmpty()) {
        Collection<Object> solrIdentifier = solrInputDoc.getFieldValues(SolrField.DC_IDENT);
        solrIdentifier.addAll(listOfCNB);

        SolrUtils.setModify(solrInputDoc, SolrField.DC_IDENT, solrIdentifier);
        SolrUtils.setModify(solrInputDoc, SolrField.MODIFIED_DATE, SolrUtils.findOutCurrentTimeStr());
}
```
Týmto, si uchovám predošlé hodnoty z DC_IDENT. Potom k ním pridám nový list, ktorý spolu so starými dátami SETnem do DC_IDENT. 
Či to nie je ono a mám to prerobiť?

kazooo commented:
pises spravne, ale *solrInputDoc* mas na zacatku uplne prazdny,
```java
final String uuidDoc = (String) solrDoc.getFieldValue(SolrField.UUID);
final SolrInputDocument solrInputDoc = new SolrInputDocument();
final Optional<Document> foxml = fedoraClient.getDC(uuidDoc);

foxml.ifPresent(doc -> fieldSynchronizers.forEach(x -> x.accept(doc, solrInputDoc)));
```
cize kus
```java
Collection<Object> solrIdentifier = solrInputDoc.getFieldValues(SolrField.DC_IDENT);
solrIdentifier.addAll(listOfCNB);
```
ti vyhodi NullPointer exception, protoze *solrInputDoc.getFieldValues(SolrField.DC_IDENT)* vrati null, a nasledne *null.addAll(listOfCNB);* nemuze byt

tady bych asi dal nasledujici konstrukce
```java
final SolrInputDocument solrInputDoc = createSolrInputDocument(solrDoc);
...
// createSolrInputDocument
final Collection<Object> existingDcIdent = solrDoc.getFieldValues(SolrField.DC_IDENT);
final SolrInputDocument inputDocument = new SolrInputDocument();
inputDocument.addField(SolrField.DC_IDENT, existingDcIdent != null ? existingDcIdent : Collections.emptyList());
return inputDocument;
```

Romx98 committed:
Modify code based PR comments

- added class contains File Write utils
- modify methods for current time date
- save existingDcIdent to InputDoc
- change the condition for adding a new solr doc
- change log when adding solr doc
- create static object to foxmlUtils
- added new dir for logs

kazooo commented on src/main/java/cz/mzk/scripts/SyncFoxmlFieldsWithSolr.java:
```diff
@@ -46,13 +49,15 @@ public static void main(String[] args) throws ParserConfigurationException, Tran
 
-            foxml.ifPresent(doc -> fieldSynchronizers.forEach(x -> x.accept(doc, solrInputDoc)));
+            foxml.ifPresentOrElse(doc -> fieldSynchronizers.forEach(x -> x.accept(doc, solrInputDoc)),
+                    () -> missingDocInFedora.writeLine(uuidDoc));
```
dobre, jen formatovani

```java
foxml.ifPresentOrElse(
    doc -> fieldSynchronizers.forEach(x -> x.accept(doc, solrInputDoc)),
    () -> missingDocInFedora.writeLine(uuidDoc)
);
```
cte se lip :)

kazooo commented on src/main/java/cz/mzk/scripts/SyncFoxmlFieldsWithSolr.java:
```diff
@@ -46,13 +49,15 @@ public static void main(String[] args) throws ParserConfigurationException, Tran
 
-            if (!solrInputDoc.isEmpty()) {
-                solrClient.sendSolrInputDocument(solrInputDoc);
+            if (solrInputDoc.containsKey(SolrField.MODIFIED_DATE)) {
```
.isEmpty() nefunguje? nebo proc rozhodujes podle MODIFIED_DATE?

Romx98 replied:
Kvoli tomu, ze uz pri vytvarani InputDocumentu obsahuje field name DC_IDENT, cize kontrola isEmpty mi pride v tomto bode zbytocna, lebo tou podmienkou prejde vsetko. 
Preto ma napadlo kontrolovat to podla MODIFIED_DATE, ktory sa pridava len do dokumentov, ktore vykonali nejaku zmenu.
Ci, som to zle pochopil a bude sa to dat kontrolovat aj s pomocou isEmpty?

kazooo commented on src/main/java/cz/mzk/scripts/SyncFoxmlFieldsWithSolr.java:
```diff
@@ -104,7 +109,13 @@ private static SolrQuery createFetchAllRootsWithoutTheRequiredFieldsSolrQuery(fi
         return solrQuery;
     }
+
+    private static SolrInputDocument createSolrInputDocument(SolrDocument solrDoc) {
```
final

kazooo commented on src/main/java/cz/mzk/utils/FileWriterUtils.java:
```diff
@@ -0,0 +1,45 @@
+            final BufferedWriter bufferedWriter = new BufferedWriter(fileWriter);
+            bufferedWriter.write(rootUuid);
+            bufferedWriter.newLine();
+            bufferedWriter.close();
```
pokazde kdyz je potreba zapsat radek, oteviras soubor, zapisujes a zaviras
je to spatne, IO operace jsou drahe a neni tento postup optimalni
BufferedWritter klidne muze zustat atributem FileWriterUtils po otevreni, po zpracovani skriptem proste uzavres zdroje

kazooo commented on src/main/java/cz/mzk/utils/FileWriterUtils.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.Date;
+
+@Slf4j
+public class FileWriterUtils {
```
Utils je spis postfix pro statickou tridu, ktera ma sadu utilit
tady mas wrapper pro BufferedWriter, Path a proste wrapper pro soubor, do ktereho zapisujes
takze je to semanticky matouci kdyz davas wrapperu postfix utils
dokonce nemusi to byt i Writter, budes pozdeji pouzivat wrapper i na cteni ze souboru

co treba FileWrapper?

Romx98 replied:
Ked si to teraz citam, tak to dava celkom zmysel :) okej, FileWrapper znie ovela lepsie

Romx98 committed:
Modify FileWrapper

- create new methods contains close buffer
- modify attributes

kazooo committed:
Add script runner configurations

kazooo committed:
Add logging and refactor some parts

kazooo committed:
Fix file wrapper and setup logging

kazooo committed:
Configure script outputs and Solr query building

kazooo committed:
Finish script

kazooo committed:
Refactor some pieces of code
